### PR TITLE
Fix segfault when closing the game before the data has finished loading.

### PR DESCRIPTION
**Pull request by quyykk**
## Fix Details

Closing the game before the game data finished loading crashed the game. The reason is that the `future` storing the async computation is a global. A future's destructor waits for the computation to finish before returning, but data loading uses various globals which may or may not already be destroyed when the destructor is ran (because the future is itself a global).

The fix is simply to make the future part of `main`, so that when it is destroyed every global still exists.

Another fix would be to cancel the data loading thread, to avoid having to load data when we know the game has been closed. This isn't possible with `async`/`future`, so we'd have to use a proper `thread` and co. like other multithreaded parts. ~~Or instead use better threading abstractions like tbb~~.

## Testing Done

I'm on Linux. Start the game and immediately close it. On master: segfault. Not with this PR though.

### CODE REVIEW

File: source/main.cpp [hunk 1]
```diff
@@ -116,10 +116,11 @@ int main(int argc, char *argv[])
...
 
+	future<void> dataLoading;
 	try {
 		// Begin loading the game data.
 		bool isConsoleOnly = loadOnly || printShips || printTests || printWeapons;
-		GameData::BeginLoad(isConsoleOnly, debugMode);
+		dataLoading = GameData::BeginLoad(isConsoleOnly, debugMode);
 
 		// If we are not using the UI, or performing some automated task, we should load
 		// all data now. (Sprites and sounds can safely be deferred.)
```
Comment (tehhowch) [inline source/main.cpp:126]: Should we be attempting to consume the future in these no-ui blocks?

File: source/main.cpp [hunk 2]
```diff
@@ -116,10 +116,11 @@ int main(int argc, char *argv[])
 	}
 	Files::Init(argv);
 
+	future<void> dataLoading;
 	try {
```
Comment (tehhowch) [inline source/main.cpp:120]: What issues arise if the future is scoped within the try block?

File: source/main.cpp (hunk 1 above)
Comment (quyykk) [reply source/main.cpp:126]: We could I guess. Should I? We'd need to call `wait_for(0s)` which looks a bit weird but not a big problem I guess.

**Comment by quyykk** [comment]
Those integration tests hate me again. It looks like they fail because SDL can't find a video device. I really don't understand why though, or how that is relevant to the last commit I made here.

**Comment by quyykk** [comment]
@tehhowch Honestly I'd prefer to revert the last commit I made (that uses the `future` to wait), because I have absolutely no clue why this causes the integration tests to fail and the errors are consistent on every run.

**Comment by tehhowch** [comment]
> @tehhowch Honestly I'd prefer to revert the last commit I made (that uses the `future` to wait), because I have absolutely no clue why this causes the integration tests to fail and the errors are consistent on every run.

Feel free to open it up. We're also missing a direct inclusion of `<future>` in main.cpp, with it currently coming transitively via the GameData.h header.

It would also be worth seeing if scoping the future's lifetime beyond the try-catch helps any.

**Pull request merged**